### ISSUE TITLE
Incorrect type on Setter

### ISSUE THREAD

**Issue by ramiel** [pmndrs/jotai#1141]
I have this code

```ts
export const focusedPeerIdAtom = atom<string | null, string>(
  null,
  (get, set, peerId) => {
    const currentPeerId = get(focusedPeerIdAtom);
    if (currentPeerId === peerId) {
      
      
      set(focusedPeerIdAtom, null);
    
    
    
    } else {
      set(focusedPeerIdAtom, peerId);
    }
  }
);
```

The idea is that the setter function act as a toggler. The problem is that null cannot be assigned

![Screenshot_20220505_085012](https://user-images.githubusercontent.com/324687/166874976-96aca353-3ddc-40e0-a8ae-0b1d73a2ee62.png)

I feel like this is wrong: `set` function should accept values of type `Value` not of type `Update`. Am I understanding it wrong?

**Comment by ramiel** [comment]
So there are two topics here: the functionality and the types. First of all, is this usage correct? Should I be able to use the setter like I did (regardless of the types). I'm going to try to have a look at the types, to see if there's a workaround or a solution

**Comment by dai-shi** [comment]
Yes, `atom(initialValue, writeFunction)` is a valid usage, but just undocumented.

**Comment by ramiel** [comment]
Does it change anything if I use the following version?
`atom(readFunction, writeFunction)`

I mean, the problem is not there right?

**Comment by ramiel** [comment]
oh wait, is my code wrong becaue the atom is trying to write itself? Is this forbidden?

**Comment by dai-shi** [comment]
Yeah. It's not forbidden, but undocumented, which means not recommended.
So, the recommended way is to use two atoms.
```js
const baseAtom = atom<string | null>(null);
export const focusedPeerIdAtom = atom(
  (get) => get(baseAtom)
  (get, set, peerId: string) => {
    const currentPeerId = get(baseAtom);
    if (currentPeerId === peerId) {
      set(basetom, null);
    } else {
      set(baseAtom, peerId);
    }
  }
);
```

---
For `atom(readFunction, writeFunction)`, if you try to write the self atom, it will throw an error.

**Comment by ramiel** [comment]
Got it, I see the point. Thank you
